### ISSUE TITLE
Understanding applyMatrix

### ISSUE THREAD

**Issue by francofusco** [bernhardfritz/piksel#3]
Hello! Great job, I really like this project!

I have an issue related to the use of `Graphics::apllyMatrix`. My goal is to call plot functions so that the origin is located at the bottom-left corner of the window, with the *x* and *y* axes pointing respectively to the right and upward. The transformation equations should be written as:
```
x = u
y = h - v
```
where `u` and `v` are *image coordinates*, *i.e.*, from the original coordinate system, while `x` and `y` correspond to the custom system. The constant `h` is the height of the image. In matrix form, this should correspond to the homogeneous transformation:
```
| x |   | 1  0  0 |   | u |
| y | = | 0 -1  h | * | v | 
| 1 |   | 0  0  1 |   | 1 |
```
However, if I call `Graphics::apllyMatrix(1, 0, 0, -1, 0, h)` I just get the reflection of the vertical axis, but without any translation. Instead, if I perform the reflection only and then call `Graphics::translate(0, -h)` (`-h` is due to the fact that the reflection has already been done) everything works fine. Could you please help me understanding why the first call (reflection+translation at the same time) does not work?

**Comment by bernhardfritz** [comment]
Hi @francofusco !

Glad to hear you like it 🙂. You are right ofc, the parameters you used for Graphics::applyMatrix are correct for what you are trying to achieve. I checked the implementation of Graphics::applyMatrix and it seems there is a bug which needs to be fixed. Thank you for reporting this, nice find!

In the meanwhile, I found a workaround that should also work for you. The matrix from your comment above would be equal to the product of a scale-matrix and a translation-matrix, see:
https://www.wolframalpha.com/input/?i=%7B%7B1,+0,+0%7D,+%7B0,+-1,+0%7D,+%7B0,+0,+1%7D%7D+.+%7B%7B1,+0,+0%7D,+%7B0,+1,+-h%7D,+%7B0,+0,+1%7D%7D

This means you could also simply scale the y-axis by -1 and translate y by -height like this:
```cpp
#include "app.hpp"

void App::setup() {}

void App::draw(piksel::Graphics& g) {
    g.background(glm::vec4(0.5f, 0.5f, 0.5f, 1.0f));
    g.scale(1.0f, -1.0f);
    g.translate(0.0f, -height);
    g.rect(50, 50, 100, 100);
}
```

**Comment by francofusco** [comment]
Thanks for the quick reply!

> I checked the implementation of Graphics::applyMatrix and it seems there is a bug which needs to be fixed.

Now that you confirmed it, I checked the implementation as well. You have probably found the problem already, but in any case it should be that the translation parameters `e` and `f` end up being part of the 3D rotation matrix in the homogenous transformation and not of the translation vector. I changed [this line](https://github.com/bernhardfritz/piksel/blob/79690f467b17f8f0a5f68ad39dbebe2598868503/src/piksel/graphics.cpp#L13) into
```
glm::mat4 m = glm::mat4(
      a, b, 0, 0,
      c, d, 0, 0,
      0, 0, 1, 0,
      e, f, 0, 1
    );
```
and everything works fine! If you want, I can send a PR for this.

In any case, I think I will use your solution (`scale(1.0f, -1.0f)` + `translate(0, -height)`) as it is actually clearer, thanks for the idea :)
